File: backend/app/schemas/category.py
# app/schemas/category.py
from __future__ import annotations

from typing import Optional, List, Dict, Any
from uuid import UUID
from datetime import datetime
from pydantic import Field, ConfigDict

from app.schemas.security_schemas import (
    EntityBase,      # code, name, description, active + validaciones
    SecureBaseModel,
    CODE_RX,         # regex seguro para "code"
)

# ---------- Base ----------
class CategoryBase(EntityBase):
    pass

# ---------- DTOs de entrada ----------
class CategoryCreate(CategoryBase):
    pass

class CategoryUpdate(CategoryBase):
    pass

class CategoryPatch(SecureBaseModel):
    code: Optional[str] = Field(
        None, min_length=1, max_length=EntityBase.CODE_MAX, pattern=CODE_RX
    )
    name: Optional[str] = Field(None, min_length=1, max_length=EntityBase.NAME_MAX)
    description: Optional[str] = Field(None, max_length=EntityBase.DESC_MAX)
    active: Optional[bool] = None

# ---------- DTOs de salida ----------
class CategoryRead(SecureBaseModel):
    id: UUID
    code: str
    name: str
    description: Optional[str] = None
    active: bool
    created_at: datetime
    updated_at: Optional[datetime] = None
    user_id: Optional[UUID] = None

    model_config = ConfigDict(from_attributes=True)

class CategoryListResponse(SecureBaseModel):
    total: int
    items: List[CategoryRead]

    model_config = ConfigDict(from_attributes=True)

# ---------- Resultado de importación ----------
class CategoryImportResult(SecureBaseModel):
    total_imported: int
    total_errors: int
    imported: List[str]
    errors: List[Dict[str, Any]]

File: frontend/app/types/category.ts
export interface Category {
  id: string;
  code: string;
  name: string;
  description?: string;
  active: boolean;
  createdAt?: string | Date;
  updatedAt?: string | Date;
  userId?: string;
}

export interface CreateCategoryData {
  code: string;
  name: string;
  description?: string;
  active?: boolean;
}

export interface UpdateCategoryData extends CreateCategoryData {
  id: string;
}

export interface CategoryListResponseDTO {
  total: number;
  items: Category[];
}

// Utilidad para mapear datos del backend al modelo Category
export function mapCategory(data: any): Category {
  return {
    id: data.id,
    code: data.code,
    name: data.name,
    description: data.description,
    active: !!data.active,
    createdAt: data.createdAt,
    updatedAt: data.updatedAt,
    userId: data.userId,
  };
}
